src/std_extensions/collections.rs

```diff
@@ -59,50 +59,60 @@ impl CustomConst for ListValue {
         crate::values::downcast_equal_consts(self, other)
     }
 }
+const TP: TypeParam = TypeParam::Type(TypeBound::Any);
 
 fn extension() -> Extension {
     let mut extension = Extension::new(EXTENSION_NAME);
 
     extension
         .add_type(
             LIST_TYPENAME,
-            vec![TypeParam::Type(TypeBound::Any)],
+            vec![TP],
             "Generic dynamically sized list of type T.".into(),
             TypeDefBound::FromParams(vec![0]),
         )
         .unwrap();
+    let temp_reg: ExtensionRegistry = [extension.clone()].into();
+    let list_type_def = extension.get_type(&LIST_TYPENAME).unwrap();
+
+    let (l, e) = list_types(list_type_def);
     extension
-        .add_op_custom_sig_simple(
+        .add_op_type_scheme_simple(
             POP_NAME,
             "Pop from back of list".into(),
-            vec![TypeParam::Type(TypeBound::Any)],
-            move |args: &[TypeArg]| {
-                let (list_type, element_type) = list_types(args)?;
-                Ok(FunctionType {
-                    input: TypeRow::from(vec![list_type.clone()]),
-                    output: TypeRow::from(vec![list_type, element_type]),
-                    extension_reqs: ExtensionSet::singleton(&EXTENSION_NAME),
-                })
-            },
+            PolyFuncType::new_validated(
+                vec![TP],
+                FunctionType::new(vec![l.clone()], vec![l.clone(), e.clone()]),
+                &temp_reg,
+            )
+            .unwrap(),
         )
         .unwrap();
     extension
-        .add_op_custom_sig_simple(
+        .add_op_type_scheme_simple(
             PUSH_NAME,
             "Push to back of list".into(),
-            vec![TypeParam::Type(TypeBound::Any)],
-            move |args: &[TypeArg]| {
-                let (list_type, element_type) = list_types(args)?;
-                Ok(FunctionType {
-                    output: TypeRow::from(vec![list_type.clone()]),
-                    input: TypeRow::from(vec![list_type, element_type]),
-                    extension_reqs: ExtensionSet::singleton(&EXTENSION_NAME),
-                })
-            },
+            PolyFuncType::new_validated(
+                vec![TP],
+                FunctionType::new(vec![l.clone(), e], vec![l]),
+                &temp_reg,
+            )
+            .unwrap(),
         )
         .unwrap();
     extension
 }
+
+fn list_types(list_type_def: &TypeDef) -> (Type, Type) {
```

Comment:
Did you mean to move this?

`list_and_elem_type` (or `list_type_and_elem`) might be a better name